lib/TPP/Transforms/CombineXsmmPass.cpp

```diff
@@ -0,0 +1,181 @@
+//
+
+#include "TPP/Dialect/Xsmm/XsmmOps.h"
+#include "TPP/Dialect/Xsmm/XsmmUtils.h"
+#include "TPP/Passes.h"
+#include "TPP/Transforms/Utils/BuilderUtils.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+using namespace mlir;
+
+#include "TPP/Dialect/Xsmm/XsmmUtils.h"
+#include "TPP/Transforms/Utils/VNNIUtils.h"
+namespace mlir {
+namespace tpp {
+#define GEN_PASS_DEF_COMBINEXSMMOPPASS
+#include "TPP/Passes.h.inc"
+} // namespace tpp
+} // namespace mlir
+namespace {
+
+static FailureOr<DenseI64ArrayAttr>
+getSizesAndLeadingDimForBrgemmOp(RewriterBase &rewriter, xsmm::BrgemmOp opTy) {
+
+  auto memrefC = opTy.getOperand(3).getType();
+  auto memrefA = opTy.getOperand(1).getType();
+  auto memrefB = opTy.getOperand(2).getType();
+
+  int64_t m, n, k;
+  if (!isa<ShapedType>(memrefC) || !isa<ShapedType>(memrefA)) {
+    return failure();
+  }
+  m = memrefC.cast<ShapedType>().getShape()[0];
+  n = memrefC.cast<ShapedType>().getShape()[1];
+  k = memrefA.cast<ShapedType>().getShape()[2];
+
+  auto ldaDim = xsmm::utils::getLeadingDim(memrefA, /*pos=*/1);
+  if (failed(ldaDim)) {
+    return failure();
+  }
+  int64_t lda = *ldaDim;
+
+  auto ldbDim = xsmm::utils::getLeadingDim(memrefB, /*pos=*/1);
+  if (failed(ldbDim)) {
+    return failure();
+  }
+  int64_t ldb = (vnni::utils::isInVnniLayout(memrefB.cast<MemRefType>()))
+                    ? *ldbDim / *vnni::utils::getVnniBlockingFactor(memrefB)
+                    : *ldbDim;
+
+  auto ldcDim = xsmm::utils::getLeadingDim(memrefC);
+  if (failed(ldcDim)) {
+    return failure();
+  }
+  int64_t ldc = *ldcDim;
+
+  // If we are dealing with a BRGEMM we need to pass two extra dimensions:
+  // - strideA and strideB that represent the stride between different GEMM
+  // in BRGEMM.
+  int64_t strideA = lda * m;
+  int64_t strideB = ldb * k;
+  return DenseI64ArrayAttr::get(
+      rewriter.getContext(),
+      ArrayRef<int64_t>{m, n, k, lda, ldb, ldc, strideA, strideB});
+}
+
+static ArrayAttr getBrgemmFlags(RewriterBase &rewriter, xsmm::BrgemmOp opTy) {
+  auto memrefB = opTy.getOperand(2).getType().cast<MemRefType>();
+  xsmm::GemmFlagsAttr gemmFlag =
+      (vnni::utils::isInVnniLayout(memrefB))
+          ? xsmm::GemmFlagsAttr::get(rewriter.getContext(),
+                                     xsmm::GemmFlags::VNNI_B)
+          : xsmm::GemmFlagsAttr::get(rewriter.getContext(),
+                                     xsmm::GemmFlags::NONE);
+  return rewriter.getArrayAttr(gemmFlag);
+}
+
+struct CombineXsmmOp : public OpRewritePattern<xsmm::BrgemmOp> {
+  using OpRewritePattern<xsmm::BrgemmOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(xsmm::BrgemmOp brgemmOp,
+                                PatternRewriter &rewriter) const override {
+    auto use = brgemmOp.getOperand(3).getDefiningOp();
+    xsmm::BinaryOp binaryOp;
+    xsmm::UnaryOp unaryOp;
+    for (auto user : use->getUsers()) {
```

Comment:
This is insufficient.

If the pattern is `GEMM + RELU + ADD`, this will match and it will fuse wrong. Not that there's any reason for the pattern to be that, but we need to write safe matchers.

There are the following cases:
* GEMM + ADD + RELU: match all three
* GEMM + ADD: match binary only
* GEM + RELU: match unary only

Everything else, for now, is not a match.

Variations we need to detect and reject:
* GEMM + RELU + ADD
* GEMM + ??? + ADD + RELU
* GEMM + ADD + ??? + RELU

Variations we don't care but still need to match:
* GEMM + ADD + RELU + ??? -> FUSED_GEMM + ???
* Zero + GEMM + ADD + RELU -> Zero + FUSED_GEMM
* Later:  Zero + FUSED_GEMM -> FUSED_GEMM(Beta=0)

There are two ways of doing this without graph structured matchers:
1. Force the order in a particular way for the users (`getUser[1].isa<Add>()` etc). This is inflexible and leads to too many if/else blocks.
2. Loop through all users (like you do), but verify that you have the right number of ops in the right order. You can match expected ops, even if you don't handle them (ex. Zero, for now). But you have to continue on everh match, and you have to stop once you already have matched all you can.

To do (2) above, you need to:
* Check for Zero, and ignore for now, `continue`.
* Check for Add, verify structure/flags/operands, save and `continue`
* Check for Relu, verify structure/flags/operands, save and `continue`
* If you have a unary (which is meant to be the last op), you must stop looking.

If you don't continue on each match, you'll be trying to match everything every time. If you don't stop at the unary, you'll try to fuse following ops that don't fuse.

In the end, you need to add the fused op and remove all matches above. Because this is in `memref`, you don't need to reconstruct the chain between the last non-match op from before and the first non-match op from after, since they'll all be operating (reading/writing) directly to the `memref` (alloc, subview).